### PR TITLE
Moved kubemark master from Debian to GCI

### PR DESCRIPTION
**Pull request by shyamjvs**
This PR fixes issue #37484
Kubemark master now runs on GCI instead of Debian, taking it one step closer to a real cluster master.
Primary changes:
1. changing master VM image/OS in kubemark's config-default.sh to debian
2. moving kubelet to systemd from supervisord
3. changing directory for cert/key/csv files from /srv/kubernetes to /etc/srv/kubernetes

cc @kubernetes/sig-scalability-misc  @wojtek-t @gmarek

### CODE REVIEW

**Comment by timothysc** [comment]
How do we get our images (Red Hat) into this mix?  

I think we could probably save a lot of time on elaborate bug hunts by bisecting across our images.

/cc @dchen1107 @jeremyeder @gmarek

**Comment by gmarek** [comment]
That's also on our plate as well.

**Comment by timothysc** [comment]
Whatever we need to do to get you subscriptions to latest RHEL, just ask. 

/cc @eparis

File: test/kubemark/resources/start-kubemark-master.sh [hunk 1]
```diff
@@ -21,6 +21,24 @@
 KUBE_ROOT="/home/kubernetes"
 KUBE_BINDIR="${KUBE_ROOT}/kubernetes/server/bin"
 
+function config-ip-firewall {
+  echo "Configuring IP firewall rules"
+  # The GCI image has host firewall which drop most inbound/forwarded packets.
```
Comment (gmarek) [inline test/kubemark/resources/start-kubemark-master.sh]: @vishh - can you tell why this seems to be needed?

**Comment by gmarek** [comment]
Looks good - I'd want to wait for comment from @vishh before giving a final lgtm.

File: test/kubemark/resources/start-kubemark-master.sh (hunk 1 above)
Comment (gmarek) [inline test/kubemark/resources/start-kubemark-master.sh]: cc @mtaufen @dchen1107, as IIUC @vishh is OOO.

Comment (mtaufen) [inline test/kubemark/resources/start-kubemark-master.sh]: I don't know off the top of my head but it looks like this hunk is straight out of `cluster/gce/gci/configure-helper.sh`. @freehan @andyzheng0831 `git blame` points to you two, any background info on this?

**Comment by shyamjvs** [comment]
Yes, it has been picked from .../gci/configure-helper.sh and it is required
because GCI seems to have tight network firewall rules by default.

On Fri, Dec 16, 2016 at 3:40 PM Michael Taufen <notifications@github.com>
wrote:

*@mtaufen* commented on this pull request.
------------------------------

In test/kubemark/resources/start-kubemark-master.sh
<https://github.com/kubernetes/kubernetes/pull/38616>:

> @@ -21,6 +21,24 @@
 KUBE_ROOT="/home/kubernetes"
 KUBE_BINDIR="${KUBE_ROOT}/kubernetes/server/bin"

+function config-ip-firewall {
+  echo "Configuring IP firewall rules"
+  # The GCI image has host firewall which drop most inbound/forwarded packets.

I don't know off the top of my head but it looks like this hunk is straight
out of cluster/gce/gci/configure-helper.sh. @freehan
<https://github.com/freehan> @andyzheng0831
<https://github.com/andyzheng0831> git blame points to you two, any
background info on this?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/38616>, or mute the thread

File: test/kubemark/resources/start-kubemark-master.sh (hunk 1 above)
Comment (freehan) [inline test/kubemark/resources/start-kubemark-master.sh]: This is needed because GCI defaults to drop all inbound packet and will not forward packet either. Hence leaving the node useless.

**Comment by shyamjvs** [comment]
@k8s-bot gci gke e2e test this

File: test/kubemark/resources/start-kubemark-master.sh (hunk 1 above)
Comment (wojtek-t) [inline test/kubemark/resources/start-kubemark-master.sh]: Thanks for clarification!

File: test/kubemark/resources/start-kubemark-master.sh [hunk 2]
```diff
@@ -123,7 +141,6 @@ function assemble-docker-flags {
 	# TODO(shyamjvs): Incorporate network plugin options, etc later.
 	echo "DOCKER_OPTS=\"${docker_opts}\"" > /etc/default/docker
 	echo "DOCKER_NOFILE=65536" >> /etc/default/docker  # For setting ulimit -n
-	service docker restart
```
Comment (wojtek-t) [inline test/kubemark/resources/start-kubemark-master.sh]: Why did you remove this in this PR?

File: test/kubemark/resources/start-kubemark-master.sh [hunk 3]
```diff
@@ -176,39 +193,43 @@ function compute-kubelet-params {
...
-	if [[ -n "${builtin_kubelet}" ]]; then
-		cp "${KUBE_BINDIR}/kubelet" "$(dirname "$builtin_kubelet")"
-	fi
+	# Create systemd config.
+	local -r kubelet_bin="/usr/bin/kubelet"
+	create-kubelet-conf "${kubelet_bin}"
 
-	# Create supervisord config for kubelet.
-	create-kubelet-conf "${KUBE_BINDIR}/kubelet"
+	# Flush iptables nat table.
```
Comment (wojtek-t) [inline test/kubemark/resources/start-kubemark-master.sh]: Why do we need this?

File: test/kubemark/resources/start-kubemark-master.sh (hunk 2 above)
Comment (shyamjvs) [inline test/kubemark/resources/start-kubemark-master.sh]: The gci configure-helper.sh wasn't restarting docker unless using a network plugin (which it was using anyway since "${NETWORK_PROVIDER:-}" == "kubenet"). Thus it restarts the docker daemon through systemctl always. However, in the case of kubemark I haven't made it restart following the gci's configure-helper.sh script. Have fixed this issue now in kubemark by restarting docker and I think that's a bug in configure-helper.sh that needs a fix (restarting even when not using network plugin).

File: test/kubemark/resources/start-kubemark-master.sh (hunk 3 above)
Comment (shyamjvs) [inline test/kubemark/resources/start-kubemark-master.sh]: This is again something that I carried on from the gci configure-helper.sh and have removed it now. The nat IP table doesn't seem to do anything much significant  (and e2e-test works both with/without any rules in this table) except for allowing IP masquerading on post-routing. If you are curious, here is how the nat table looks like before flushing:

Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere             anywhere             ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
DOCKER     all  --  anywhere            !127.0.0.0/8          ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
KUBE-POSTROUTING  all  --  anywhere             anywhere             /* kubernetes postrouting rules */
MASQUERADE  all  --  172.17.0.0/16        anywhere            

Chain DOCKER (2 references)
target     prot opt source               destination         
RETURN     all  --  anywhere             anywhere            

Chain KUBE-MARK-DROP (0 references)
target     prot opt source               destination         
MARK       all  --  anywhere             anywhere             MARK or 0x8000

Chain KUBE-MARK-MASQ (0 references)
target     prot opt source               destination         
MARK       all  --  anywhere             anywhere             MARK or 0x4000

Chain KUBE-POSTROUTING (1 references)
target     prot opt source               destination         
MASQUERADE  all  --  anywhere             anywhere             /* kubernetes service traffic requiring SNAT */ mark match 0x4000/0x4000

**Comment by wojtek-t** [comment]
LGTM - will merge tomorrow early morning to monitor if nothing will break.

File: test/kubemark/resources/start-kubemark-master.sh (hunk 1 above)
Comment (vishh) [inline test/kubemark/resources/start-kubemark-master.sh]: +1 to what @freehan mentioned!
